Docs Infra:Apply Jupyter caching mechanism

# Quarto Jupyter Cache: Comprehensive Issue Tracking & Enhancement Proposal

> **Status**: Open  
> **Labels**: `cache`, `jupyter`, `parallel-rendering`, `reproducibility`, `enhancement`  
> **Related Projects**: [`quarto-cli`](https://github.com/quarto-dev/quarto-cli), [`jupyter-cache`](https://github.com/executablebooks/jupyter-cache)  
> **Last Updated**: 2026-01-22  

---

## 1. Executive Summary

Quarto's Jupyter execution cache (`execute.cache: true`) is designed to cache document-level execution results. While beneficial for reducing redundant computation, the current implementation introduces several limitations that impact reproducibility, developer workflow, and CI/CD reliability—particularly in multi-file research projects or collaborative environments.

This document consolidates known issues, reproduction steps, workarounds, and proposed enhancements to guide development prioritization and community discussion.

---

## 2. Identified Issues

### 2.1 Per-Cell Cache Control Not Respected (`#| cache: false` Ignored)

| Field | Details |
|-------|---------|
| **Description** | When `execute.cache: true` is set at the document level, per-cell directives like `#| cache: false` are ignored; all cells are cached regardless of individual settings. |
| **Root Cause** | Jupyter Cache operates at the notebook/document level. Per-cell cache granularity is currently only supported in the Knitr engine. |
| **Impact** | Cells relying on external state (e.g., `%run` scripts, random seeds, `datetime.now()`) incorrectly reuse stale cached outputs, compromising reproducibility. |
| **Reproduction** | See [Section 4.1](#41-minimal-reproducible-example-per-cell-cache). |
| **Related Issues** | [executablebooks/jupyter-cache#56](https://github.com/executablebooks/jupyter-cache/issues/56) |

---

### 2.2 SQLite Race Condition During Parallel Rendering

| Field | Details |
|-------|---------|
| **Error Message** | `(sqlite3.OperationalError) table settings already exists` |
| **Trigger Conditions** | - `_quarto.yml` has `project.execute-parallel: true` (default)<br>- Multiple `.qmd` files rendered concurrently (e.g., via `quarto render` on a project directory or parallel CI jobs) |
| **Root Cause** | Concurrent processes attempt to initialize the same SQLite database (`.jupyter-cache/cache.db`) simultaneously, leading to DDL race conditions during `settings` table creation. |
| **Impact** | Intermittent render failures; partial output generation; non-deterministic CI behavior. |
| **Reproduction** | See [Section 4.2](#42-minimal-reproducible-example-parallel-rendering). |

---

### 2.3 External Script Change Detection Limitations

| Field | Details |
|-------|---------|
| **Description** | Modifications to externally sourced Python files (e.g., via `%run script.py` or `import module`) are not detected by the cache invalidation logic. |
| **Current Behavior** | Cache keys are derived solely from the cell's source code string; file dependencies are not hashed or tracked. |
| **Workaround Attempts** | - `%autoreload` extension: reloads modules in-kernel but does not invalidate cache<br>- Wrapping execution in `%%bash` or `subprocess.run()`: bypasses cache but loses in-kernel state sharing |
| **Impact** | Developers must manually clear cache or disable caching to ensure external changes propagate, reducing cache utility. |

---

### 2.4 `pre-render` Hook Limitations

| Field | Details |
|-------|---------|
| **Scope Isolation** | Scripts executed via `pre-render` run in a separate process; generated artifacts cannot be directly injected into the Python execution context of the document. |
| **VS Code Preview** | `pre-render` hooks are not triggered in live preview unless `quarto preview --render all` is used. |
| **Configuration Constraint** | `pre-render` must be defined in `_quarto.yml` under `project:`; not supported in document YAML frontmatter. |
| **Impact** | Limits use cases for dynamic metadata injection, version stamping, or pre-processing pipelines. |

---

### 2.5 Cache Directory Management & CI/CD Friction

| Field | Details |
|-------|---------|
| **Default Behavior** | Cache stored in `.jupyter-cache/` relative to each `.qmd` file; no centralized configuration. |
| **CI Challenges** | - Cache size grows unbounded without manual pruning<br>- `actions/cache` `restore-keys` help but do not prevent SQLite conflicts in parallel jobs<br>- No native support for environment-variable-driven cache location (e.g., `JUPYTER_CACHE_DIR`) in stable Quarto releases |
| **Impact** | Increased CI runtime, storage costs, and flakiness in distributed build environments. |

---

## 3. Desired Behavior (Requirements)

1. ✅ **Per-cell cache precedence**: `#| cache: false` should override document-level `cache: true`.
2. ✅ **Parallel-render safety**: Concurrent renders should not cause SQLite DDL conflicts (e.g., via file locking, per-process DB shards, or deferred initialization).
3. ✅ **Dependency-aware cache keys**: Provide built-in helpers to include file hashes in cache keys (e.g., `{{ file_hash("_include/utils.py") }}` or programmatic API).
4. ✅ **`pre-render` → execution context bridging**: Allow `pre-render` outputs to be loaded as variables in Python cells (e.g., via `context: pre` or shared JSON/YAML artifacts).
5. ✅ **CI-friendly cache configuration**: Support `JUPYTER_CACHE_DIR` env var and provide cache pruning utilities (`quarto cache clean`).

---

## 4. Minimal Reproducible Examples

### 4.1 Per-Cell Cache (`cache: false` Ignored)

**File**: `test.qmd`
```yaml
---
execute:
  cache: true
---
```

````markdown
```{python}
#| cache: false
import datetime
print(f"Executed at: {datetime.datetime.now()}")
```
````

**Commands**:
```bash
quarto render test.qmd  # First run: prints timestamp A
quarto render test.qmd  # Second run: expected timestamp B, but reuses cached output A
```

---

### 4.2 Parallel Rendering SQLite Race

**Project Structure**:
```
project/
├── _quarto.yml          # execute-parallel: true (default)
├── doc1.qmd
├── doc2.qmd
└── ...
```

**Commands**:
```bash
# Trigger concurrent renders (e.g., via GNU parallel or CI matrix)
parallel quarto render ::: doc*.qmd
# OR
quarto render  # renders all files in parallel by default
```

**Observed Error**:
```
[WARNING] SQLite error: table settings already exists
[ERROR] Render failed for doc2.qmd
```

---

## 5. Current Workarounds (Community-Sourced)

| Issue | Workaround | Trade-offs |
|-------|------------|------------|
| Per-cell cache | Switch to `engine: knitr` + `reticulate` | Requires R; Python/R scope separation; ecosystem friction |
| SQLite race | Set `execute-parallel: false` in `_quarto.yml` | Linear rendering; 2–5× longer build times |
| External script changes | Use `subprocess.run(["python", "script.py"])` + file I/O | Loses in-kernel state; verbose; error handling complexity |
| `pre-render` isolation | Use `quarto render --render all` or `renderOnSave: true` in VS Code | Ties workflow to CLI flags or editor config |
| CI cache bloat | Combine `actions/cache` with periodic `--no-cache` renders + manual cleanup scripts | Imperfect; requires maintenance; not deterministic |

---

## 6. Environment Context

| Component | Version / Details |
|-----------|------------------|
| Quarto CLI | 1.6.x (tested), 1.7-beta (partial mitigation) |
| Jupyter Cache | ≥1.0.0 |
| Python Kernel | ipykernel 6.x, Python 3.10+ |
| OS | Ubuntu 22.04 (CI), macOS 14 (dev), Docker (amd64/arm64) |
| Editor | VS Code + Quarto extension v1.85+ |
| CI Platform | GitHub Actions (`ubuntu-latest`) |

---

## 7. Proposed Next Steps

### For Quarto Core Team
- [ ] Audit Jupyter Cache integration for per-cell directive support
- [ ] Implement SQLite initialization safeguards for parallel access (e.g., `CREATE TABLE IF NOT EXISTS` + retry logic)
- [ ] Expose cache key customization API (YAML + programmatic)
- [ ] Document `pre-render` → execution context data flow patterns

### For Community / Contributors
- [ ] Prototype `file_hash()` template function for cache keys
- [ ] Share CI-optimized cache strategies (e.g., cache partitioning by branch/PR)
- [ ] Test Quarto 1.7+ beta features for cache directory configuration

### For Users (Immediate Mitigations)
```yaml
# _quarto.yml
project:
  execute-parallel: false  # Disable parallel rendering to avoid SQLite races
  render:
    cache: 
      enabled: true
      # Future: cache-dir: ${JUPYTER_CACHE_DIR:-.jupyter-cache}
```

```bash
# Local dev: force re-execution of specific cells
quarto render --no-cache path/to/file.qmd

# CI: periodic full rebuild to prevent cache drift
if [ $((GITHUB_RUN_NUMBER % 20)) -eq 0 ]; then
  quarto render --no-cache
else
  quarto render
fi
```

---

## 8. References & Related Discussions

- [executablebooks/jupyter-cache#56: Per-cell cache control](https://github.com/executablebooks/jupyter-cache/issues/56)
- [quarto-dev/quarto-cli#4321: Cache invalidation on external file changes](https://github.com/quarto-dev/quarto-cli/issues/4321)
- [Quarto Docs: Execution Caching](https://quarto.org/docs/computations/execution-options.html#caching)
- [Executable Books: MyST-NB Cache Design](https://myst-nb.readthedocs.io/en/latest/authoring/cache.html)

--- a/.github/workflows/deploy-docs-ghpage.yml
+++ b/.github/workflows/deploy-docs-ghpage.yml
@@ -63,6 +63,14 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-cached-
 
+      - name: Cache _jupyter_cache directory
+        uses: actions/cache@v4
+        with:
+          path: _jupyter_cache
+          key: ${{ runner.os }}-jupyter-${{ steps.cache-key.outputs.key }}
+          restore-keys: |
+            ${{ runner.os }}-jupyter-
+
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v3
 

--- a/docs/_include/_generate_metadata_tex.py
+++ b/docs/_include/_generate_metadata_tex.py
@@ -48,7 +48,7 @@ def main():
                         help='Path to the main QMD file (e.g., index.qmd). If not provided, uses QUARTO_PROJECT_INPUT_FILE or first valid .qmd in directory.')
     parser.add_argument('--output', '-o', default='./_include/_metadata.tex',
                         help='Output LaTeX metadata file path')
-    parser.add_argument('--version_prefix', '-p', default='0.1',
+    parser.add_argument('--version_prefix', '-p', default=None,
                         help='Version prefix (e.g., 0.1)')
     parser.add_argument('--version_mark', action='store_true',
                         help='Include background version watermark in PDF')
@@ -90,8 +90,11 @@ def main():
     # ----- Compute version -----
     with open(qmd_path, 'rb') as f:
         file_hash = hashlib.sha256(f.read()).hexdigest()
-    date_short = datetime.now().strftime("%y%m%d")
-    version_str = f"{args.version_prefix}-{date_short}-{file_hash[:6]}"
+    date_short = datetime.now().strftime("%y%m%d")    
+    if args.version_prefix is not None:
+        version_str = f"{args.version_prefix}-{date_short}-{file_hash[:6]}"
+    else:
+        version_str = f"{date_short}-{file_hash[:6]}"
 
     # ----- Extract YAML front matter (author/affiliations are optional) -----
     front = extract_front_matter(qmd_path)

--- a/docs/_quarto.yml
+++ b/docs/_quarto.yml
@@ -1,5 +1,7 @@
 # Shared
+project: 
+  pre-render: python3 _quarto_pre-render.py
 csl: _include/ieee.csl
 link-citations: true
 execute:   
-  cache: false # true: should be activatied only for experimental-purpose.
+  cache: true

--- a/docs/_quarto_pre-render.py
+++ b/docs/_quarto_pre-render.py
@@ -11,14 +11,14 @@ def get_front_matter(qmd_path):
     return yaml.safe_load(m.group(1)) or {}
 
 def main():
-    input_files = os.environ.get("QUARTO_PROJECT_INPUT_FILES", "").splitlines()
+    # List of files to render at pre‑render time (here only one)ROJECT_INPUT_FILES", "").splitlines()
     qmd_files = [f for f in input_files if f.endswith('.qmd')]
     if not qmd_files:
         print("No .qmd files to process")
         return
 
-    # 프로젝트 루트에서 _generate_metadata_tex.py 찾기
-    project_root = Path(os.environ.get("QUARTO_PROJECT_ROOT", ".")).resolve()
+    # Find _generate_metadata_tex.py in your project root
+    # Project root (use if necessary)h(os.environ.get("QUARTO_PROJECT_ROOT", ".")).resolve()
     generator = project_root / "_include" / "_generate_metadata_tex.py"
 
     for qmd in qmd_files:
@@ -26,9 +26,9 @@ def main():
         front = get_front_matter(qmd_path)
         
         vpre = front.get('version-prefix', None)
-        vmark = front.get('version-mark', False)
+        # YAML fields: version-prefix (None if none), version-mark (default false)
         
-        # QMD 파일이 있는 디렉토리 아래의 _files 폴더 사용
+        # Use the _files folder under the directory where the QMD files are located.
         doc_dir = qmd_path.parent
         out_dir = doc_dir / "_files"
         out_dir.mkdir(exist_ok=True)

--- a/docs/_quarto_pre-render.py
+++ b/docs/_quarto_pre-render.py
@@ -1,0 +1,47 @@
+#!/usr/bin/env python3
+import os, sys, subprocess, re, yaml
+from pathlib import Path
+
+def get_front_matter(qmd_path):
+    with open(qmd_path, encoding='utf-8') as f:
+        content = f.read()
+    m = re.match(r'^---\s*\n(.*?)\n---\s*\n', content, re.DOTALL)
+    if not m:
+        return {}
+    return yaml.safe_load(m.group(1)) or {}
+
+def main():
+    input_files = os.environ.get("QUARTO_PROJECT_INPUT_FILES", "").splitlines()
+    qmd_files = [f for f in input_files if f.endswith('.qmd')]
+    if not qmd_files:
+        print("No .qmd files to process")
+        return
+
+    # 프로젝트 루트에서 _generate_metadata_tex.py 찾기
+    project_root = Path(os.environ.get("QUARTO_PROJECT_ROOT", ".")).resolve()
+    generator = project_root / "_include" / "_generate_metadata_tex.py"
+
+    for qmd in qmd_files:
+        qmd_path = Path(qmd).resolve()
+        front = get_front_matter(qmd_path)
+        
+        vpre = front.get('version-prefix', None)
+        vmark = front.get('version-mark', False)
+        
+        # QMD 파일이 있는 디렉토리 아래의 _files 폴더 사용
+        doc_dir = qmd_path.parent
+        out_dir = doc_dir / "_files"
+        out_dir.mkdir(exist_ok=True)
+        out_file = out_dir / f"{qmd_path.stem}_metadata.tex"
+        
+        cmd = [sys.executable, str(generator), "--input", str(qmd_path), "--output", str(out_file)]
+        if vpre is not None:
+            cmd += ["--version_prefix", str(vpre)]
+        if vmark:
+            cmd.append("--version_mark")
+        
+        print(f"Generating {out_file} from {qmd_path.name}")
+        subprocess.run(cmd, check=True)
+
+if __name__ == "__main__":
+    main()

--- a/docs/_quarto_pre-render.py
+++ b/docs/_quarto_pre-render.py
@@ -10,36 +10,50 @@ def get_front_matter(qmd_path):
         return {}
     return yaml.safe_load(m.group(1)) or {}
 
+# Return the path inside \input{...} if it ends with '_metadata.tex', else None.
+def find_input_metadata_line(qmd_path):
+    with open(qmd_path, encoding='utf-8') as f:
+        for line in f:
+            m = re.search(r'\\input\{([^}]*_metadata\.tex)\}', line)
+            if m:
+                return m.group(1)
+    return None
+
 def main():
-    # List of files to render at pre‑render time (here only one)ROJECT_INPUT_FILES", "").splitlines()
+    input_files = os.environ.get("QUARTO_PROJECT_INPUT_FILES", "").splitlines()
     qmd_files = [f for f in input_files if f.endswith('.qmd')]
     if not qmd_files:
         print("No .qmd files to process")
         return
 
-    # Find _generate_metadata_tex.py in your project root
-    # Project root (use if necessary)h(os.environ.get("QUARTO_PROJECT_ROOT", ".")).resolve()
+    project_root = Path(os.environ.get("QUARTO_PROJECT_ROOT", ".")).resolve()
     generator = project_root / "_include" / "_generate_metadata_tex.py"
 
     for qmd in qmd_files:
         qmd_path = Path(qmd).resolve()
+
+        # 1. Find the \input{..._metadata.tex} pattern
+        target_rel = find_input_metadata_line(qmd_path)
+        if not target_rel:
+            print(f"Skipping {qmd_path.name}: no \\input{{..._metadata.tex}} line found.")
+            continue
+
+        # 2. Output file path (relative to QMD's directory)
+        out_file = (qmd_path.parent / target_rel).resolve()
+        out_file.parent.mkdir(parents=True, exist_ok=True)
+
+        # 3. Read front matter
         front = get_front_matter(qmd_path)
-        
         vpre = front.get('version-prefix', None)
-        # YAML fields: version-prefix (None if none), version-mark (default false)
-        
-        # Use the _files folder under the directory where the QMD files are located.
-        doc_dir = qmd_path.parent
-        out_dir = doc_dir / "_files"
-        out_dir.mkdir(exist_ok=True)
-        out_file = out_dir / f"{qmd_path.stem}_metadata.tex"
-        
+        vmark = front.get('version-mark', False)
+
+        # 4. Build command – only add --version_prefix if vpre is not None
         cmd = [sys.executable, str(generator), "--input", str(qmd_path), "--output", str(out_file)]
         if vpre is not None:
             cmd += ["--version_prefix", str(vpre)]
         if vmark:
             cmd.append("--version_mark")
-        
+
         print(f"Generating {out_file} from {qmd_path.name}")
         subprocess.run(cmd, check=True)
 

--- a/docs/legal/index.qmd
+++ b/docs/legal/index.qmd
@@ -11,6 +11,8 @@ author:
 date: 2026-02
 date-format: "MMMM, YYYY"
 date-modified: last-modified
+version-prefix: 1.0
+version-mark: true
 metadata-files:
   - ../_include/author.yml
   - ../_include/format.html.yml
@@ -33,13 +35,6 @@ format:
             }
             \makeatother
 ---
-
-```{python}
-#| echo: false
-#| include: false
-#| context: local
-%run ../_include/_generate_metadata_tex.py --output ./_files/_metadata.tex --version_prefix 1.0 --version_mark
-```
 
 # SSCCS Foundation {.unnumbered .unlisted}
 

--- a/docs/legal/index.qmd
+++ b/docs/legal/index.qmd
@@ -26,7 +26,7 @@ format:
             \usepackage{microtype}
             \usepackage{unicode-math}
             \addtokomafont{disposition}{\rmfamily}
-            \input{./_files/_metadata.tex}
+            \input{./_files/index_metadata.tex}
             \makeatletter
             \def\@maketitle{
               \centering

--- a/docs/philosophy/foundation.qmd
+++ b/docs/philosophy/foundation.qmd
@@ -22,7 +22,7 @@ format:
   pdf: 
     include-in-header:
       text: |
-        \input{./_files/_metadata.tex} 
+        \input{./_files/foundation_metadata.tex} 
         \addtokomafont{disposition}{\rmfamily}     
         \makeatletter
         \def\@maketitle{

--- a/docs/philosophy/foundation.qmd
+++ b/docs/philosophy/foundation.qmd
@@ -47,14 +47,6 @@ abstract: |
 
 ---
 
-```{python}
-#| echo: false
-#| include: false
-#| context: local 
-
-%run ../_include/_generate_metadata_tex.py --output ./_files/_metadata.tex
-```
-
 ## The Beginning of a Question
 
 When we discuss technology, we often ask is it possible? Yet this question contains a deeper inquiry.

--- a/docs/proposal/proposal.qmd
+++ b/docs/proposal/proposal.qmd
@@ -11,6 +11,8 @@ author:
 date: 2026-03
 date-format: "MMMM, YYYY"
 date-modified: last-modified
+version-prefix: 1.0
+version-mark: true
 bibliography: ./_include/references.bib
 colorlinks: true
 metadata-files:
@@ -77,8 +79,6 @@ title_meta_items = {
 ```{python}
 #| echo: false
 #| include: false
-%run ../_include/_generate_metadata_tex.py --output ./_files/proposal_metadata.tex --version_prefix 1.0 --version_mark
-
 import matplotlib.pyplot as plt
 import numpy as np
 ```

--- a/docs/proposal/pt.qmd
+++ b/docs/proposal/pt.qmd
@@ -6,6 +6,7 @@ email: "proposal@ssccs.org"
 website: "https://ssccs.org"
 doi: "10.5281/zenodo.18759106"
 date: last-modified
+version-prefix: 1.0
 metadata-files:
   - ../_include/author.yml
   - ../_include/format.beamer.yml  
@@ -39,7 +40,6 @@ format:
 #| include: false
 #| context: local
 %run ../_include/_graphviz.py
-%run ../_include/_generate_metadata_tex.py --output ./_files/pt_metadata.tex --version_prefix 1.0
 ```
 
 ## Introducing

--- a/docs/research/riscv.qmd
+++ b/docs/research/riscv.qmd
@@ -3,6 +3,8 @@ title: "RISC-V Integration Research"
 subtitle: "Structural Observation Paradigm Validation on RISC-V Ecosystem"
 date: 2026-04
 date-format: "MMMM, YYYY"
+version-mark: true
+version-prefix: 0.1
 bibliography: ./_include/riscv_references.bib
 link-citations: true
 colorlinks: true
@@ -31,7 +33,6 @@ abstract: |
   This research report explores the integration of SSCCS (Schema–Segment Composition Computing System) across the RISC‑V ecosystem, using the OpenHW Foundation as a central validation platform. SSCCS redefines computation as structural observation rather than instruction sequencing, addressing the von Neumann bottleneck at the logical layer. Through the RISC‑V eXtension Interface (XIF) and other custom instruction mechanisms, we investigate implementing observation primitives as coprocessor extensions, evaluating energy efficiency and verifiability characteristics on real silicon. The report outlines a research pathway spanning software emulation, hardware prototyping, eFPGA integration, and community contribution, positioning SSCCS as open public infrastructure for sustainable, accountable computing across the RISC‑V landscape.
 ---
 
-
 ```{python}
 #| echo: false
 #| include: false
@@ -48,10 +49,6 @@ title_meta_items = {
 }
 
 %run ../_include/_graphviz.py
-%run ../_include/_generate_metadata_tex.py --output ./_files/riscv_metadata.tex --version_prefix 0.1 --version_mark
-
-import matplotlib
-matplotlib.use('Agg')
 ```
 
 {{< include ../_include/_title_meta_items_if_html.qmd >}}

--- a/docs/research/riscv_space.qmd
+++ b/docs/research/riscv_space.qmd
@@ -6,6 +6,8 @@ maturity-level: conceptual-exploration
 disclaimer: "This report explores hypothetical synergies and does not constitute a development specification."
 date: 2026-04
 date-format: "MMMM, YYYY"
+version-mark: true
+version-prefix: 0.2
 bibliography: ./_include/riscv_space_references.bib
 link-citations: true
 colorlinks: true
@@ -50,10 +52,6 @@ title_meta_items = {
 }
 
 %run ../_include/_graphviz.py
-%run ../_include/_generate_metadata_tex.py --output ./_files/riscv_space_metadata.tex --version_prefix 0.2 --version_mark
-
-import matplotlib
-matplotlib.use('Agg')
 ```
 
 {{< include ../_include/_title_meta_items_if_html.qmd >}}

--- a/docs/research/rust_baremetal.qmd
+++ b/docs/research/rust_baremetal.qmd
@@ -3,6 +3,8 @@ title: "Rust Bare‑Metal Toolchain as a Foundation for RISC-V Integration"
 subtitle: "Leveraging Embedded Rust and Comprehensive Rust Bare‑Metal Guides for Structural Observation Primitives"
 date: 2026-04
 date-format: "MMMM, YYYY"
+version-prefix: 0.1
+version-mark: true
 bibliography: ./_include/riscv_references.bib
 link-citations: true
 colorlinks: true
@@ -44,10 +46,6 @@ title_meta_items = {
 }
 
 %run ../_include/_graphviz.py
-%run ../_include/_generate_metadata_tex.py --output ./_files/rust_baremetal_metadata.tex --version_prefix 0.1 --version_mark
-
-import matplotlib
-matplotlib.use('Agg')
 ```
 
 {{< include ../_include/_title_meta_items_if_html.qmd >}}

--- a/docs/whitepaper/whitepaper.qmd
+++ b/docs/whitepaper/whitepaper.qmd
@@ -12,6 +12,8 @@ author:
         url: https://ssccs.org
 date: 2026-02
 date-format: "MMMM, YYYY"
+version-prefix: 0.1
+version-mark: true
 bibliography: ./_include/references.bib
 csl: ./../_include/ieee.csl
 date-modified: last-modified
@@ -68,10 +70,7 @@ format:
 #| include: false
 #| context: local
 %run ../_include/_graphviz.py
-%run ../_include/_generate_metadata_tex.py --output ./_files/_metadata.tex --version_prefix 0.1 --version_mark
-```
 
-```{python}
 title_meta_items = {
    "html": [
       {"title": "Other Formats", "link": "./whitepaper.pdf", "content": "PDF", "content_class": "bi bi-file-pdf"},

--- a/docs/whitepaper/whitepaper.qmd
+++ b/docs/whitepaper/whitepaper.qmd
@@ -45,7 +45,7 @@ format:
         \usepackage{microtype}
         \setkeys{Gin}{width=\linewidth,height=\textheight,keepaspectratio}
         \addtokomafont{disposition}{\rmfamily}
-        \input{./_files/_metadata.tex} 
+        \input{./_files/whitepaper_metadata.tex} 
         \DeclareTOCStyleEntry[beforeskip=6pt]{default}{section}
         \DeclareTOCStyleEntry[beforeskip=0pt]{default}{subsection}               
         \makeatletter

